pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -223,48 +231,52 @@ func StartDNSProxy(address string, port uint16, lookupEPFunc LookupEndpointIDByI
 	return p, nil
 }
 
-// AddAllowed adds reStr, a regexp, to the DNS lookups the proxy allows.
-func (p *DNSProxy) AddAllowed(reStr, endpointID string) {
-	log.WithField(logfields.DNSName, reStr).Debug("Adding allowed DNS FQDN pattern")
-	p.UpdateAllowed([]string{reStr}, nil, endpointID)
-}
-
-// RemoveAllowed removes reStr from the DNS lookups the proxy allows. It must
-// match the form in AddAllowed exactly (i.e. this isn't removing by regex, but
-// by direct equivalence).
-func (p *DNSProxy) RemoveAllowed(reStr, endpointID string) {
-	log.WithField(logfields.DNSName, reStr).Debug("Removing allowed DNS FQDN pattern")
-	p.UpdateAllowed(nil, []string{reStr}, endpointID)
-}
-
 // UpdateAllowed adds and removes reStr while holding the lock. This is a bit
 // of a hack to ensure atomic updates of rules until we replace the tracking
 // with something better.
-func (p *DNSProxy) UpdateAllowed(reStrToAdd, reStrToRemove []string, endpointID string) {
-	for i := range reStrToAdd {
-		reStrToAdd[i] = prepareNameMatch(reStrToAdd[i])
-	}
-	for i := range reStrToRemove {
-		reStrToRemove[i] = prepareNameMatch(reStrToRemove[i])
-	}
-
+func (p *DNSProxy) UpdateAllowed(endpointID uint64, newRules policy.L7DataMap) {
 	p.Lock()
 	defer p.Unlock()
-	for _, reStr := range reStrToRemove {
-		p.allowed.Remove(reStr, endpointID)
-	}
-	for _, reStr := range reStrToAdd {
-		p.allowed.Add(reStr, endpointID)
-	}
+	// TODO: Add regex precompiles
+	p.allowedByID[endpointID] = newRules
 }
 
 // CheckAllowed checks name against the rules added to the proxy, and only
 // returns true if this endpointID was added (via AddAllowed) previously.
-func (p *DNSProxy) CheckAllowed(name, endpointID string) bool {
-	name = strings.ToLower(name)
+func (p *DNSProxy) CheckAllowed(endpointID uint64, destID identity.NumericIdentity, name string) bool {
```

Comment:
Is this sufficiently abstracted to unit-test it? Seems like there are unit tests for some cases already, might be able to expand them to test the L7 behaviour?

Reply:
Yes, very much so. In fact, it might be simple to unit test the various rule combos now, instead of doing it in the e2e tests. I'll expand on the tests we have.